File: frappe/core/page/background_jobs/background_jobs.py
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
# License: MIT. See LICENSE

from typing import TYPE_CHECKING

import frappe
from frappe.utils import convert_utc_to_user_timezone
from frappe.utils.background_jobs import get_queues, get_workers
from frappe.utils.scheduler import is_scheduler_inactive

if TYPE_CHECKING:
	from rq.job import Job

JOB_COLORS = {"queued": "orange", "failed": "red", "started": "blue", "finished": "green"}


@frappe.whitelist()
def get_info(view=None, queue_timeout=None, job_status=None) -> list[dict]:
	jobs = []

	def add_job(job: "Job", queue: str) -> None:

		if job.kwargs.get("site") == frappe.local.site:
			job_info = {
				"job_name": job.kwargs.get("kwargs", {}).get("playbook_method")
				or job.kwargs.get("kwargs", {}).get("job_type")
				or str(job.kwargs.get("job_name")),
				"status": job.get_status(),
				"queue": queue,
				"creation": convert_utc_to_user_timezone(job.created_at),
				"color": JOB_COLORS[job.get_status()],
			}

			if job.exc_info:
				job_info["exc_info"] = job.exc_info

			jobs.append(job_info)

	if view == "Jobs":
		queues = get_queues()
		for queue in queues:
			for job in queue.jobs:
				if job_status != "all" and job.get_status() != job_status:
					return
				if queue_timeout != "all" and not queue.name.endswith(f":{queue_timeout}"):
					return
				add_job(job, queue.name)

	elif view == "Workers":
		workers = get_workers()
		for worker in workers:
			current_job = worker.get_current_job()
			if current_job:
				if hasattr(current_job, "kwargs") and current_job.kwargs.get("site") == frappe.local.site:
					add_job(current_job, current_job.origin)
				else:
					jobs.append({"queue": worker.name, "job_name": "busy", "status": "", "creation": ""})
			else:
				jobs.append({"queue": worker.name, "job_name": "idle", "status": "", "creation": ""})

	return jobs


@frappe.whitelist()
def remove_failed_jobs():
	queues = get_queues()
	for queue in queues:
		fail_registry = queue.failed_job_registry
		for job_id in fail_registry.get_job_ids():
			job = queue.fetch_job(job_id)
			fail_registry.remove(job, delete_job=True)


@frappe.whitelist()
def get_scheduler_status():
	if is_scheduler_inactive():
		return {"status": "inactive"}
	return {"status": "active"}

File: frappe/core/page/background_jobs/background_jobs.js
frappe.pages["background_jobs"].on_page_load = (wrapper) => {
	const background_job = new BackgroundJobs(wrapper);

	$(wrapper).bind("show", () => {
		background_job.show();
	});

	window.background_jobs = background_job;
};

class BackgroundJobs {
	constructor(wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: __("Background Jobs"),
			single_column: true,
		});

		this.page.add_inner_button(__("Remove Failed Jobs"), () => {
			frappe.confirm(__("Are you sure you want to remove all failed jobs?"), () => {
				frappe
					.call("frappe.core.page.background_jobs.background_jobs.remove_failed_jobs")
					.then(() => this.refresh_jobs());
			});
		});

		this.page.main.addClass("frappe-card");
		this.page.body.append('<div class="table-area"></div>');
		this.$content = $(this.page.body).find(".table-area");

		this.make_filters();
		this.refresh_jobs = frappe.utils.throttle(this.refresh_jobs.bind(this), 1000);
	}

	make_filters() {
		this.view = this.page.add_field({
			label: __("View"),
			fieldname: "view",
			fieldtype: "Select",
			options: ["Jobs", "Workers"],
			default: "Jobs",
			change: () => {
				this.queue_timeout.toggle(this.view.get_value() === "Jobs");
				this.job_status.toggle(this.view.get_value() === "Jobs");
			},
		});
		this.queue_timeout = this.page.add_field({
			label: __("Queue"),
			fieldname: "queue_timeout",
			fieldtype: "Select",
			options: [
				{ label: "All Queues", value: "all" },
				{ label: "Default", value: "default" },
				{ label: "Short", value: "short" },
				{ label: "Long", value: "long" },
			],
			default: "all",
		});
		this.job_status = this.page.add_field({
			label: __("Job Status"),
			fieldname: "job_status",
			fieldtype: "Select",
			options: [
				{ label: "All Jobs", value: "all" },
				{ label: "Queued", value: "queued" },
				{ label: "Deferred", value: "deferred" },
				{ label: "Started", value: "started" },
				{ label: "Finished", value: "finished" },
				{ label: "Failed", value: "failed" },
			],
			default: "all",
		});
		this.auto_refresh = this.page.add_field({
			label: __("Auto Refresh"),
			fieldname: "auto_refresh",
			fieldtype: "Check",
			default: 1,
			change: () => {
				if (this.auto_refresh.get_value()) {
					this.refresh_jobs();
				}
			},
		});
	}

	show() {
		this.refresh_jobs();
		this.update_scheduler_status();
	}

	update_scheduler_status() {
		frappe.call({
			method: "frappe.core.page.background_jobs.background_jobs.get_scheduler_status",
			callback: (r) => {
				let { status } = r.message;
				if (status === "active") {
					this.page.set_indicator(__("Scheduler: Active"), "green");
				} else {
					this.page.set_indicator(__("Scheduler: Inactive"), "red");
				}
			},
		});
	}

	refresh_jobs() {
		let view = this.view.get_value();
		let args;
		let { queue_timeout, job_status } = this.page.get_form_values();
		if (view === "Jobs") {
			args = { view, queue_timeout, job_status };
		} else {
			args = { view };
		}

		this.page.add_inner_message(__("Refreshing..."));
		frappe.call({
			method: "frappe.core.page.background_jobs.background_jobs.get_info",
			args,
			callback: (res) => {
				this.page.add_inner_message("");

				let template = view === "Jobs" ? "background_jobs" : "background_workers";
				this.$content.html(
					frappe.render_template(template, {
						jobs: res.message || [],
					})
				);

				let auto_refresh = this.auto_refresh.get_value();
				if (frappe.get_route()[0] === "background_jobs" && auto_refresh) {
					setTimeout(() => this.refresh_jobs(), 2000);
				}
			},
		});
	}
}
